added export list to Operations, refactored message handling.

wygulmage opened the pull request:
### Description

The current definition of `broadcastMessage` seems to be O(n^2) in the number of workspaces because it uses `sendMessageWithNoRefresh` and `sendMessageWithNoRefresh` uses `updateLayout` and `updateLayout` uses `runOnWorkspaces`.

This changes `broadCastMessage` and `sendMessageWithNoRefresh` to each use a single call to `runOnWorkspaces`.

Because it uses two new helper functions, I also added an explicit export list to `XMonad.Operations`. The export list defines sections as indicated by comments in the source code. (`XMonad.Operations` could use a documentation cleanup, but this is not that.)

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I've confirmed these changes don't belong in xmonad-contrib instead

  - [ ] I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

  - [ ] I updated the `CHANGES.md` file

slotThe commented on src/XMonad/Operations.hs:
```diff
@@ -403,17 +430,24 @@ sendMessage a = windowBracket_ $ do
+broadcastMessage = filterMessageWithNoRefresh (const True)
 
 -- | Send a message to a layout, without refreshing.
 sendMessageWithNoRefresh :: Message a => a -> W.Workspace WorkspaceId (Layout Window) Window -> X ()
```
This may be a good time to change this to `Message a => a -> WindowSpace -> X ()` for more clarity.

liskin commented:
This looks nice, but there's one concern I have: `runOnWorkspaces` runs the monadic action on the hidden workspaces first, then current and visible, whereas `broadcastMessage` did it in current, visible, hidden order. I'm not entirely sure it's safe to change this order, as some may depend on it. :-/
(Opinions from others welcome.)

Other than that, the accompanying text that you wrote in the pull request description should go into the commit message as well. Also, whenever you feel like using comma in the commit message subject, that's a hint you might consider splitting it into two. ;-)
More at https://chris.beams.io/posts/git-commit/, https://who-t.blogspot.com/2009/12/on-commit-messages.html

ghost commented:
> Also, whenever you feel like using comma in the commit message subject, that's a hint you might consider splitting it into two. ;-)
> More at https://chris.beams.io/posts/git-commit/, https://who-t.blogspot.com/2009/12/on-commit-messages.html

Good call. I'll split this up into 3 PRs: one for the export list, one to cleanup documentation comments, and one to do the actual refactor I want.

liskin commented:
> Good call. I'll split this up into 3 PRs: one for the export list, one to cleanup documentation comments, and one to do the actual refactor I want.

Oh no please don't. Commits, not PRs! Please reopen this one, there's absolutely no need to add more noise with additional PRs.

Commit:
added export list to Operations.

This includes Haddock sections that will make the documentation render differently.

Commit:
updated documentation of Operations.

Added missing documentation for type alias `D`.

Moved misplaced documentation.

Edited to a consistent style that will play well with Haddock.

Commit:
changed `broadcastMessage` to be O(workspaces).

The current definition of broadcastMessage seems to be O(n^2) in the number of workspaces because it uses sendMessageWithNoRefresh and sendMessageWithNoRefresh uses updateLayout and updateLayout uses runOnWorkspaces.

This changes broadCastMessage and sendMessageWithNoRefresh to each use a single
call to runOnWorkspaces.

ghost commented:
> > Good call. I'll split this up into 3 PRs: one for the export list, one to cleanup documentation comments, and one to do the actual refactor I want.
> 
> Oh no please don't. Commits, not PRs! Please reopen this one, there's absolutely no need to add more noise with additional PRs.

Got it. Sorry about the confusion! Three separate commits now. Should be clear what they do. This goes a bit farther than the previous commits because it fixes up the documentation a bit too. Things looked really odd with correct sections but incorrect Haddock comments.

Now I just need to deal with order-of-operations issues. For now I'm thinking of writing a version of `runOnWorkspaces` that uses the same order as the original `broadcastMessage`.

Commit:
changed order of `broadcastMessage` to match original.

This uses two new functions (not exported):
`workspacesA` traverses the workspaces in a StackSet.
`runOnWorkspaces` runs `workspacesA` with the X state.

ghost commented:
OK, the order of operations is as before. To do this I added two more non-exported functions. One traverses the workspaces in a stackset in the correct order, the other one does `runOnWorkspaces` in the right order. Both might be useful to export from somewhere (probably XMonad.StackSet for the traversal), but that's probably out of scope for this PR.

slotThe commented on src/XMonad/Operations.hs:
```diff
@@ -190,8 +191,9 @@ windowBracket p action = withWindowSet $ \old -> do
--- @X@ action reporting its need for refresh via @Any@.
+-- | Perform an @X@ action. If it returns @Any True@, unwind the changes to the @WindowSet@ and replay them using @windows@.
+-- This is a version of @windowBracket@ that discards the return value and
+-- handles an @X@ action that reports its need for refresh via @Any@.
```
I think it's worth trying to be consistent with the doc-style of the rest of the module here (and also in other parts); meaning line-breaks after what I assume are 71 or 79 columns.

slotThe commented on src/XMonad/Operations.hs:
```diff
@@ -14,7 +14,33 @@
 -----------------------------------------------------------------------------
 
-module XMonad.Operations where
+module XMonad.Operations (
```
This is really cool :+1:

slotThe commented on src/XMonad/Operations.hs:
```diff
@@ -431,17 +431,24 @@ sendMessage a = windowBracket_ $ do
 sendMessageWithNoRefresh a w =
-    handleMessage (W.layout w) (SomeMessage a) `catchX` return Nothing >>=
-    updateLayout  (W.tag w)
+    filterMessageWithNoRefresh (on (==) W.tag w) a
```
I don't know if this is just me, but I always find `on` to be terribly confusing if it's not written in an infix style, because when it reads like "do something on `(==)`" instead of "do `(==)` on something`

slotThe commented on src/XMonad/Operations.hs:
```diff
@@ -496,6 +497,31 @@ initColor dpy c = C.handle (\(C.SomeException _) -> return Nothing) $
+workspacesA ::
+    (Applicative m)=>
+    (W.Workspace i l a -> m (W.Workspace i' l' a)) ->
+    W.StackSet i l a sid sd -> m (W.StackSet i' l' a sid sd)
```
This formatting (ditto `runOnWorkspaces'`) looks kind of funny

geekosaur replied:
Once you convert a branch on your repo into a pull request, any xmonad maintainer has access to the PR. Of course, you're asking for trouble with force pushes in that case.

slotThe replied:
You can tick the "allow access from maintainers" checkbox somewhere in the PR, but I think it is on by default.  I think the grouping is quite understandable; I was just thrown off a bit due to the trailing arrow syntax, as opposed to something like 

``` haskell
workspacesA
    :: Applicative m
    => (W.Workspace i l a -> m (W.Workspace i' l' a))
    -> W.StackSet i l a sid sd -> m (W.StackSet i' l' a sid sd)
```

liskin commented:
Just got back to this and I noticed `runOnWorkspaces` isn't used anywhere else than in `updateLayout`, so it might be sensible to just change it instead of introducing `runOnWorkspaces'`. Nothing in xmonad-contrib uses it either, so we don't need to think about whether something depends on the order or not—it's simply not used at all. Possibly a very small number of users have this in their config, but a changelog entry will deal with this.

Sorry for the long delay. No need to rush making the change, either, I'll be happy to take over and finish it and merge later next week.

ghost commented:
No worries about the delay. I agree that there's absolutely no rush. It's a change that, hopefully, does not change anything except for speeding up message handling when you have a lot of workspaces.

Please proceed with as much caution as you think is necessary. I haven't run into any problems, but better to have consistent n^2 message handling than someone's complex setup breaking because messages arrive at layouts in a slightly different order.

I've cleaned up the formatting per @slotThe 's helpful suggestions and rebased.

Commit:
generalized signature of `updateLayoutsBy`

liskin commented:
> Please proceed with as much caution as you think is necessary.

Just a heads up what my plan here is: I'll change the `runOnWorkspaces` order and thus drop some of the new functions in X.Core. I'm quite confident it's safe enough to do. I meant to do that yesterday but spent most of the day packing my backpack instead. I'll get back to this next wednesday/thursday. No action needed, I'll finish this.

liskin committed:
X.Operations: Small cleanup of module header

liskin committed:
Revert "generalized signature of `updateLayoutsBy`"

This reverts commit eb48bb4aef638ba6958a4489a282c274e9779767.

liskin committed:
Revert "changed order of `broadcastMessage` to match original."

We'll adjust runOnWorkspaces instead.

This reverts commit 5eff329fc607624f843f7b29fb5a4f2169e92cd0.

liskin committed:
Rename updateLayoutsBy to modifyLayouts

More consistent with our existing naming.

liskin committed:
Adjust runOnWorkspaces processing order

We now use this in `broadcastMessage`, so to not change which workspaces
get the message first, we need to change the order here. This wouldn't
normally be safe to do either, but there are no other uses of
`runOnWorkspaces` neither here nor in xmonad-contrib, so it should
actually be fine.

liskin commented:
Not exactly "next wednesday/thursday" but it's done! Thanks. :-)
